File: src/signlab/extraction/tracking.py
"""Deterministic two-hand identity tracking over vendor-neutral detections."""

from __future__ import annotations

import math
from dataclasses import dataclass
from itertools import combinations, permutations

from signlab.extraction.types import (
    HandDetection,
    HandTrackingResult,
    LandmarkPoint,
    TrackedHand,
)

_SLOT_IDS = ("hand_0", "hand_1")
_PALM_LANDMARK_INDICES = (0, 5, 9, 13, 17)


@dataclass(frozen=True, slots=True)
class HandTrackingConfig:
    """Versionable algorithm facts used to decide whether identities continue.

    These values govern identity assignment only. They are not dataset-quality
    thresholds and do not reject or transform source observations.
    """

    max_spatial_cost: float = 0.25
    handedness_disagreement_penalty: float = 0.05
    ambiguity_margin: float = 1e-9

    def __post_init__(self) -> None:
        for field_name, value in (
            ("max_spatial_cost", self.max_spatial_cost),
            ("handedness_disagreement_penalty", self.handedness_disagreement_penalty),
            ("ambiguity_margin", self.ambiguity_margin),
        ):
            if not math.isfinite(value) or value < 0.0:
                raise ValueError(f"{field_name} must be finite and non-negative")


@dataclass(frozen=True, slots=True)
class _Assignment:
    pairs: tuple[tuple[int, int], ...]
    cost: float
    signature: tuple[tuple[str, tuple[float | str | int, ...]], ...]


def _centroid(points: tuple[LandmarkPoint, ...]) -> tuple[float, float, float]:
    count = len(points)
    return (
        sum(point.x for point in points) / count,
        sum(point.y for point in points) / count,
        sum(point.z for point in points) / count,
    )


def _palm_centroid(detection: HandDetection) -> tuple[float, float, float]:
    return _centroid(tuple(detection.image_landmarks[index] for index in _PALM_LANDMARK_INDICES))


def _squared_distance(
    first: tuple[float, float, float],
    second: tuple[float, float, float],
) -> float:
    return sum((left - right) ** 2 for left, right in zip(first, second, strict=True))


def _spatial_cost(previous: HandDetection, current: HandDetection) -> float:
    previous_wrist = previous.image_landmarks[0]
    current_wrist = current.image_landmarks[0]
    return _squared_distance(
        (previous_wrist.x, previous_wrist.y, previous_wrist.z),
        (current_wrist.x, current_wrist.y, current_wrist.z),
    ) + _squared_distance(_palm_centroid(previous), _palm_centroid(current))


def _assignment_cost(
    previous: HandDetection,
    current: HandDetection,
    config: HandTrackingConfig,
) -> tuple[float, float]:
    spatial = _spatial_cost(previous, current)
    disagrees = (
        previous.reported_handedness is not None
        and current.reported_handedness is not None
        and previous.reported_handedness != current.reported_handedness
    )
    total = spatial + (config.handedness_disagreement_penalty if disagrees else 0.0)
    return spatial, total


def _detection_sort_key(detection: HandDetection) -> tuple[float | str | int, ...]:
    """Order by untouched image geometry, then lexical and detector-index facts."""

    wrist = detection.image_landmarks[0]
    palm = _palm_centroid(detection)
    return (
        wrist.x,
        wrist.y,
        wrist.z,
        palm[0],
        palm[1],
        palm[2],
        detection.reported_handedness or "",
        detection.detector_index,
    )


def _enumerate_assignments(
    previous: tuple[HandDetection | None, HandDetection | None],
    detections: tuple[HandDetection, ...],
    config: HandTrackingConfig,
) -> tuple[_Assignment, ...]:
    active_slots = tuple(index for index, detection in enumerate(previous) if detection is not None)
    assignments: list[_Assignment] = []
    for pair_count in range(min(len(active_slots), len(detections)) + 1):
        for selected_slots in combinations(active_slots, pair_count):
            for selected_detections in permutations(range(len(detections)), pair_count):
                pairs = tuple(zip(selected_slots, selected_detections, strict=True))
                costs: list[float] = []
                valid = True
                for slot_index, detection_position in pairs:
                    prior = previous[slot_index]
                    assert prior is not None
                    spatial, total = _assignment_cost(
                        prior,
                        detections[detection_position],
                        config,
                    )
                    if spatial > config.max_spatial_cost:
                        valid = False
                        break
                    costs.append(total)
                if not valid:
                    continue
                signature = tuple(
                    (_SLOT_IDS[slot_index], _detection_sort_key(detections[detection_position]))
                    for slot_index, detection_position in pairs
                )
                assignments.append(
                    _Assignment(pairs=pairs, cost=math.fsum(costs), signature=signature)
                )
    return tuple(assignments)


def _unambiguous_pairs(
    assignments: tuple[_Assignment, ...],
    config: HandTrackingConfig,
) -> tuple[tuple[int, int], ...]:
    """Keep only pairs shared by every competitively optimal assignment."""

    max_pair_count = max(len(assignment.pairs) for assignment in assignments)
    fullest = tuple(
        assignment for assignment in assignments if len(assignment.pairs) == max_pair_count
    )
    ordered = sorted(fullest, key=lambda assignment: (assignment.cost, assignment.signature))
    best_cost = ordered[0].cost
    competitive = tuple(
        assignment
        for assignment in ordered
        if assignment.cost <= best_cost + config.ambiguity_margin
    )
    common_pairs = set(competitive[0].pairs)
    for assignment in competitive[1:]:
        common_pairs.intersection_update(assignment.pairs)
    return tuple(sorted(common_pairs))


def _fallback_pairs(
    previous: tuple[HandDetection | None, HandDetection | None],
    detections: tuple[HandDetection, ...],
    available_slots: tuple[int, ...],
    detection_positions: tuple[int, ...],
    config: HandTrackingConfig,
) -> tuple[tuple[int, int], ...]:
    """Place every unmatched observation without claiming confident continuity."""

    candidates: list[
        tuple[
            int,
            float,
            tuple[tuple[str, tuple[float | str | int, ...]], ...],
            tuple[tuple[int, int], ...],
        ]
    ] = []
    for selected_slots in combinations(available_slots, len(detection_positions)):
        for ordered_positions in permutations(detection_positions):
            pairs = tuple(zip(selected_slots, ordered_positions, strict=True))
            overwritten_slots = sum(previous[slot_index] is not None for slot_index, _ in pairs)
            costs = []
            for slot_index, detection_position in pairs:
                prior = previous[slot_index]
                if prior is not None:
                    _, total = _assignment_cost(prior, detections[detection_position], config)
                    costs.append(total)
            signature = tuple(
                (_SLOT_IDS[slot_index], _detection_sort_key(detections[detection_position]))
                for slot_index, detection_position in pairs
            )
            candidates.append((overwritten_slots, math.fsum(costs), signature, pairs))
    return min(candidates)[-1]


def _result(detections: list[HandDetection | None]) -> HandTrackingResult:
    return HandTrackingResult(
        slots=(
            TrackedHand(slot_id="hand_0", detection=detections[0]),
            TrackedHand(slot_id="hand_1", detection=detections[1]),
        )
    )


class HandIdentityTracker:
    """Assign at most two detections to stable, deterministic output slots.

    Initialization uses unmodified image-space geometry rather than detector order.
    Subsequent frames exhaustively compare the possible continuity assignments.
    Unambiguous in-range pairs retain their slots. Every remaining observation is
    still emitted through a deterministic geometry/index fallback; tracking
    thresholds never become observation-retention or quality policy.
    """

    def __init__(self, config: HandTrackingConfig | None = None) -> None:
        self.config = config or HandTrackingConfig()
        self._previous: tuple[HandDetection | None, HandDetection | None] = (None, None)

    def reset(self) -> None:
        """Forget prior identities so the next frame initializes from geometry."""

        self._previous = (None, None)

    def track(self, detections: tuple[HandDetection, ...]) -> HandTrackingResult:
        """Track one frame and return both ordered slots, including absences."""

        if len(detections) > 2:
            raise ValueError("HandIdentityTracker accepts at most two detections per frame")
        detector_indices = tuple(detection.detector_index for detection in detections)
        if len(set(detector_indices)) != len(detector_indices):
            raise ValueError("detector_index values must be unique within a frame")

        if not detections:
            # A detector can miss every hand for an individual frame. Emit the
            # absence without ending identities that belong to this recording;
            # the recording boundary calls ``reset`` explicitly.
            return _result([None, None])

        active_slots = tuple(
            index for index, detection in enumerate(self._previous) if detection is not None
        )
        if not active_slots:
            initialized: list[HandDetection | None] = [None, None]
            for slot_index, detection in enumerate(sorted(detections, key=_detection_sort_key)):
                initialized[slot_index] = detection
            self._previous = (initialized[0], initialized[1])
            return _result(initialized)

        assignments = _enumerate_assignments(self._previous, detections, self.config)
        pairs = _unambiguous_pairs(assignments, self.config)
        tracked: list[HandDetection | None] = [None, None]
        next_previous = list(self._previous)
        used_positions: set[int] = set()
        used_slots: set[int] = set()
        for slot_index, detection_position in pairs:
            detection = detections[detection_position]
            tracked[slot_index] = detection
            next_previous[slot_index] = detection
            used_positions.add(detection_position)
            used_slots.add(slot_index)

        fallback_positions = tuple(
            position for position in range(len(detections)) if position not in used_positions
        )
        fallback_slots = tuple(index for index in range(2) if index not in used_slots)
        for slot_index, detection_position in _fallback_pairs(
            self._previous,
            detections,
            fallback_slots,
            fallback_positions,
            self.config,
        ):
            detection = detections[detection_position]
            tracked[slot_index] = detection
            next_previous[slot_index] = detection

        self._previous = (next_previous[0], next_previous[1])
        return _result(tracked)


__all__ = ["HandIdentityTracker", "HandTrackingConfig"]

File: apps/web/src/landmarks/tracking.ts
import type { LandmarkPoint, ReportedHandedness } from "./protocol";

const PALM_LANDMARK_INDICES = [0, 5, 9, 13, 17] as const;

export const HAND_TRACKING_CONFIG = {
  maxSpatialCost: 0.25,
  handednessDisagreementPenalty: 0.05,
  ambiguityMargin: 1e-9,
} as const;

export interface HandDetection {
  readonly detectorIndex: number;
  readonly imageLandmarks: readonly LandmarkPoint[];
  readonly worldLandmarks: readonly LandmarkPoint[];
  readonly reportedHandedness: ReportedHandedness;
  readonly handednessConfidence: number;
}

export type TrackedDetections = readonly [HandDetection | null, HandDetection | null];

interface Assignment {
  readonly pairs: readonly (readonly [number, number])[];
  readonly cost: number;
}

function squaredDistance(
  first: readonly [number, number, number],
  second: readonly [number, number, number],
): number {
  const x = first[0] - second[0];
  const y = first[1] - second[1];
  const z = first[2] - second[2];
  return x * x + y * y + z * z;
}

function pointCoordinates(point: LandmarkPoint): readonly [number, number, number] {
  return [point.x, point.y, point.z];
}

function palmCentroid(detection: HandDetection): readonly [number, number, number] {
  let x = 0;
  let y = 0;
  let z = 0;
  for (const index of PALM_LANDMARK_INDICES) {
    const point = detection.imageLandmarks[index];
    if (point === undefined) {
      throw new Error("A hand detection must contain exactly 21 image landmarks");
    }
    x += point.x;
    y += point.y;
    z += point.z;
  }
  const count = PALM_LANDMARK_INDICES.length;
  return [x / count, y / count, z / count];
}

function spatialCost(previous: HandDetection, current: HandDetection): number {
  const previousWrist = previous.imageLandmarks[0];
  const currentWrist = current.imageLandmarks[0];
  if (previousWrist === undefined || currentWrist === undefined) {
    throw new Error("A hand detection must contain exactly 21 image landmarks");
  }
  return (
    squaredDistance(pointCoordinates(previousWrist), pointCoordinates(currentWrist)) +
    squaredDistance(palmCentroid(previous), palmCentroid(current))
  );
}

function assignmentCost(
  previous: HandDetection,
  current: HandDetection,
): readonly [number, number] {
  const spatial = spatialCost(previous, current);
  const handednessPenalty =
    previous.reportedHandedness === current.reportedHandedness
      ? 0
      : HAND_TRACKING_CONFIG.handednessDisagreementPenalty;
  return [spatial, spatial + handednessPenalty];
}

function compareNumber(left: number, right: number): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

function compareString(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

function compareDetections(left: HandDetection, right: HandDetection): number {
  const leftWrist = left.imageLandmarks[0];
  const rightWrist = right.imageLandmarks[0];
  if (leftWrist === undefined || rightWrist === undefined) {
    throw new Error("A hand detection must contain exactly 21 image landmarks");
  }
  const leftPalm = palmCentroid(left);
  const rightPalm = palmCentroid(right);
  const numericComparisons = [
    compareNumber(leftWrist.x, rightWrist.x),
    compareNumber(leftWrist.y, rightWrist.y),
    compareNumber(leftWrist.z, rightWrist.z),
    compareNumber(leftPalm[0], rightPalm[0]),
    compareNumber(leftPalm[1], rightPalm[1]),
    compareNumber(leftPalm[2], rightPalm[2]),
  ];
  for (const comparison of numericComparisons) {
    if (comparison !== 0) return comparison;
  }
  const handednessComparison = compareString(left.reportedHandedness, right.reportedHandedness);
  return handednessComparison === 0
    ? compareNumber(left.detectorIndex, right.detectorIndex)
    : handednessComparison;
}

function comparePairSignatures(
  left: readonly (readonly [number, number])[],
  right: readonly (readonly [number, number])[],
  detections: readonly HandDetection[],
): number {
  for (let index = 0; index < Math.min(left.length, right.length); index += 1) {
    const leftPair = left[index];
    const rightPair = right[index];
    if (leftPair === undefined || rightPair === undefined) continue;
    const slotComparison = compareNumber(leftPair[0], rightPair[0]);
    if (slotComparison !== 0) return slotComparison;
    const leftDetection = detections[leftPair[1]];
    const rightDetection = detections[rightPair[1]];
    if (leftDetection === undefined || rightDetection === undefined) {
      throw new Error("Tracking assignment referenced a missing detection");
    }
    const detectionComparison = compareDetections(leftDetection, rightDetection);
    if (detectionComparison !== 0) return detectionComparison;
  }
  return compareNumber(left.length, right.length);
}

function combinations(values: readonly number[], count: number): number[][] {
  if (count === 0) return [[]];
  const output: number[][] = [];
  values.forEach((value, index) => {
    for (const remainder of combinations(values.slice(index + 1), count - 1)) {
      output.push([value, ...remainder]);
    }
  });
  return output;
}

function permutations(values: readonly number[], count: number): number[][] {
  if (count === 0) return [[]];
  const output: number[][] = [];
  values.forEach((value, index) => {
    const remaining = [...values.slice(0, index), ...values.slice(index + 1)];
    for (const remainder of permutations(remaining, count - 1)) {
      output.push([value, ...remainder]);
    }
  });
  return output;
}

function enumerateAssignments(
  previous: TrackedDetections,
  detections: readonly HandDetection[],
): Assignment[] {
  const activeSlots = [0, 1].filter((index) => previous[index] !== null);
  const positions = detections.map((_, index) => index);
  const assignments: Assignment[] = [];
  const maximumPairs = Math.min(activeSlots.length, detections.length);

  for (let pairCount = 0; pairCount <= maximumPairs; pairCount += 1) {
    for (const slots of combinations(activeSlots, pairCount)) {
      for (const orderedPositions of permutations(positions, pairCount)) {
        const pairs = slots.map((slot, index) => [slot, orderedPositions[index] ?? -1] as const);
        let cost = 0;
        let valid = true;
        for (const [slot, position] of pairs) {
          const prior = previous[slot];
          const current = detections[position];
          if (prior === null || prior === undefined || current === undefined) {
            valid = false;
            break;
          }
          const [spatial, total] = assignmentCost(prior, current);
          if (spatial > HAND_TRACKING_CONFIG.maxSpatialCost) {
            valid = false;
            break;
          }
          cost += total;
        }
        if (valid) assignments.push({ pairs, cost });
      }
    }
  }
  return assignments;
}

function unambiguousPairs(
  assignments: readonly Assignment[],
  detections: readonly HandDetection[],
): readonly (readonly [number, number])[] {
  const maximumPairCount = Math.max(...assignments.map(({ pairs }) => pairs.length));
  const fullest = assignments.filter(({ pairs }) => pairs.length === maximumPairCount);
  fullest.sort(
    (left, right) =>
      compareNumber(left.cost, right.cost) ||
      comparePairSignatures(left.pairs, right.pairs, detections),
  );
  const best = fullest[0];
  if (best === undefined) return [];
  const competitive = fullest.filter(
    ({ cost }) => cost <= best.cost + HAND_TRACKING_CONFIG.ambiguityMargin,
  );
  return best.pairs.filter(([slot, position]) =>
    competitive.every(({ pairs }) =>
      pairs.some(([candidateSlot, candidatePosition]) => {
        return slot === candidateSlot && position === candidatePosition;
      }),
    ),
  );
}

function fallbackPairs(
  previous: TrackedDetections,
  detections: readonly HandDetection[],
  availableSlots: readonly number[],
  detectionPositions: readonly number[],
): readonly (readonly [number, number])[] {
  const candidates: Array<{
    overwrittenSlots: number;
    cost: number;
    pairs: readonly (readonly [number, number])[];
  }> = [];
  for (const slots of combinations(availableSlots, detectionPositions.length)) {
    for (const orderedPositions of permutations(detectionPositions, detectionPositions.length)) {
      const pairs = slots.map((slot, index) => [slot, orderedPositions[index] ?? -1] as const);
      let overwrittenSlots = 0;
      let cost = 0;
      for (const [slot, position] of pairs) {
        const prior = previous[slot];
        const current = detections[position];
        if (prior !== null && prior !== undefined && current !== undefined) {
          overwrittenSlots += 1;
          cost += assignmentCost(prior, current)[1];
        }
      }
      candidates.push({ overwrittenSlots, cost, pairs });
    }
  }
  candidates.sort(
    (left, right) =>
      compareNumber(left.overwrittenSlots, right.overwrittenSlots) ||
      compareNumber(left.cost, right.cost) ||
      comparePairSignatures(left.pairs, right.pairs, detections),
  );
  return candidates[0]?.pairs ?? [];
}

export class HandIdentityTracker {
  private previous: TrackedDetections = [null, null];

  reset(): void {
    this.previous = [null, null];
  }

  track(detections: readonly HandDetection[]): TrackedDetections {
    if (detections.length > 2) {
      throw new Error("HandIdentityTracker accepts at most two detections per frame");
    }
    const detectorIndices = detections.map(({ detectorIndex }) => detectorIndex);
    if (new Set(detectorIndices).size !== detectorIndices.length) {
      throw new Error("detectorIndex values must be unique within a frame");
    }
    if (detections.length === 0) return [null, null];

    const activeSlots = [0, 1].filter((index) => this.previous[index] !== null);
    if (activeSlots.length === 0) {
      const ordered = [...detections].sort(compareDetections);
      const initialized: TrackedDetections = [ordered[0] ?? null, ordered[1] ?? null];
      this.previous = initialized;
      return initialized;
    }

    const pairs = unambiguousPairs(enumerateAssignments(this.previous, detections), detections);
    const tracked: [HandDetection | null, HandDetection | null] = [null, null];
    const nextPrevious: [HandDetection | null, HandDetection | null] = [...this.previous];
    const usedSlots = new Set<number>();
    const usedPositions = new Set<number>();

    for (const [slot, position] of pairs) {
      const detection = detections[position];
      if (detection === undefined || (slot !== 0 && slot !== 1)) continue;
      tracked[slot] = detection;
      nextPrevious[slot] = detection;
      usedSlots.add(slot);
      usedPositions.add(position);
    }

    const fallbackPositions = detections
      .map((_, index) => index)
      .filter((index) => !usedPositions.has(index));
    const fallbackSlots = [0, 1].filter((index) => !usedSlots.has(index));
    for (const [slot, position] of fallbackPairs(
      this.previous,
      detections,
      fallbackSlots,
      fallbackPositions,
    )) {
      const detection = detections[position];
      if (detection === undefined || (slot !== 0 && slot !== 1)) continue;
      tracked[slot] = detection;
      nextPrevious[slot] = detection;
    }

    this.previous = nextPrevious;
    return tracked;
  }
}
